[Editor] Implement the new alt text flow (bug 1909604)

calixteman opened the pull request:
For the Firefox pdf viewer, we want to use AI to guess an alt-text when adding an image to a pdf. For now the telemtry stuff is not implemented and will come soon.
In order to test it locally:
 - set enableAltText, enableFakeMLManager and enableUpdatedAddImage to true.
or in Firefox:
 - set browser.ml.enable, pdfjs.enableAltText and pdfjs.enableUpdatedAddImage to true.

calixteman committed:
Implement the new alt text flow (bug 1909604)

For the Firefox pdf viewer, we want to use AI to guess an alt-text when adding an image to a pdf.
For now the telemtry stuff is not implemented and will come soon.
In order to test it locally:
 - set enableAltText, enableFakeMLManager and enableUpdatedAddImage to true.
or in Firefox:
 - set browser.ml.enable, pdfjs.enableAltText and pdfjs.enableUpdatedAddImage to true.

calixteman commented:
/botio-linux preview

flodolo commented on l10n/en-US/viewer.ftl:
```diff
@@ -416,3 +416,52 @@ pdfjs-editor-colorpicker-red =
+#   $percent (Number) - the percentage of the downloaded size.
+pdfjs-editor-new-alt-text-ai-model-downloading-progress =
+    .aria-valuemin = 0
+    .aria-valuemax = $totalSize
```
```suggestion
    .aria-valuemax = { $totalSize }
```

Variables in Fluent are surrounded by `{ }`(also applies to all the other variables in these strings)

flodolo commented on l10n/en-US/viewer.ftl:
```diff
@@ -416,3 +416,52 @@ pdfjs-editor-colorpicker-red =
+    .aria-valuemin = 0
+    .aria-valuemax = $totalSize
+    .aria-valuenow = $downloadedSize
+    .aria-valuetext = Downloading alt text AI model ($downloadedSize of $totalSize Mb)
```
Is it really bits and not bytes?

flodolo commented on src/display/editor/alt_text.js:
```diff
@@ -87,6 +109,35 @@ class AltText {
+  serialize(isForCopying) {
+    let altText = this.#altText;
+    if (!isForCopying && this.guessedText === altText) {
+      altText = this.#aiPrefix.replace("{ $generatedAltText }", altText);
```
This doesn't feel very safe. Can you use `data-l10n-attrs` to pass the variable?

calixteman replied:
I did that because the generated string (from l10n engine) was containing some markers which are not in the ascii range.
So instead of workarounding, I fixed the bug:
https://github.com/mozilla/pdf.js/pull/18493

Snuffleupagus commented:
Given the sheer size of this PR, it'll take some time for me to properly review all of this. Leaving some high level observations and questions for now:
 - Since we have an upcoming release (maybe this weekend), this should *not* land just before that anyway.
 - There's been some changes/updates of the PR since it's opened, and I'd like to know that the patch has "stabilized" before doing a full review.
 - These changes mean that we now have *two separate* implementations for the altText-handling in the UI, what's the eventual plan for consolidating this functionality?
   Will the new dialog be usable in the GENERIC viewer, but obviously with AI generation switched off there, since it'd seem quite unfortunate if we'd need to maintain two separate implementations going forward?
 - The `FakeMLManager`-code should probably be limited, using the pre-processor, to just development- and testing-mode.
 - Looking at the preview above it no longer seems possible to mark an image as `decorative`, is that intended?
 - There seems to be a little bit more leading padding before the icon in the inline toolbar now, is this intended since it leads to more empty space there (see the arrow in the screen-shot below)?
    
   ![padding](https://github.com/user-attachments/assets/4953d944-aa3c-4da8-bef1-f70592ef3b0e)

calixteman commented:
> Given the sheer size of this PR, it'll take some time for me to properly review all of this. Leaving some high level observations and questions for now:
> 
>     * Since we have an upcoming release (maybe this weekend), this should _not_ land just before that anyway.

The soft freeze in m-c is next week (thursday) and we really want it asap.
The integration tests are passing locally so hopefully nothing is broken.
That said if it causes a problem for the release, I'd prefer we postpone the release itself.

> 
>     * There's been some changes/updates of the PR since it's opened, and I'd like to know that the patch has "stabilized" before doing a full review.

It should be stabilized now.

> 
>     * These changes mean that we now have _two separate_ implementations for the altText-handling in the UI, what's the eventual plan for consolidating this functionality?

We want to experiment the new flow + the ai stuff in the next release and we'll decide what we want to do then.

>       Will the new dialog be usable in the GENERIC viewer, but obviously with AI generation switched off there, since it'd seem quite unfortunate if we'd need to maintain two separate implementations going forward?

Whatever we decide, I'll do the cleanup, but for now we must live with I guess. But for sure, we (/I) don't want to have a duplicated flow...

> 
>     * The `FakeMLManager`-code should probably be limited, using the pre-processor, to just development- and testing-mode.

Fixed.

> 
>     * Looking at the preview above it no longer seems possible to mark an image as `decorative`, is that intended?

Yes it's. 

> 
>     * Will the COMPONENTS-build work correctly with the new image-resources added here, note [this code](https://github.com/mozilla/pdf.js/blob/2efa3e460c94ad6ad48541e2ec44af6e22153bda/gulpfile.mjs#L1095-L1102)?

Fixed.

> 
>     * There seems to be a little bit more leading padding before the icon in the inline toolbar now, is this intended since it leads to more empty space there (see the arrow in the screen-shot below)?

Fixed.

Snuffleupagus commented:
> The soft freeze in m-c is next week (thursday) and we really want it asap.

Well, I hope you understand that I cannot review well over a thousand lines of new code immediately (and I'm still not at a 100% unfortunately).
Especially since a lot of this is stuff is asynchronous, which requires more time to consider how the various pieces interact.

One quickly observed bug, related to the asynchronicity of the code:
 1. Open the preview in https://github.com/mozilla/pdf.js/pull/18492#issuecomment-2247947991.
 2. Add a new Stamp-editor.
 3. Immediately close the altText dialog *and* delete the editor.

In the console the following is printed (after a short wait):
```
TypeError: can't access property "guessedAltText", this[#currentEditor] is null
    #mlGuessAltText new_alt_text_manager.js:248
new_alt_text_manager.js:257:14
```

calixteman commented:
> > The soft freeze in m-c is next week (thursday) and we really want it asap.
> 
> Well, I hope you understand that I cannot review well over a thousand lines of new code immediately (and I'm still not at a 100% unfortunately). Especially since a lot of this is stuff is asynchronous, which requires more time to consider how the various pieces interact.
>

I completely understand your point, but we want to make the experiment in 130.
If you see something obviously wrong I'll fix it immediately, but if you don't notice any bad things, we could merge it and fix the issues we may discover (especially when I'll begin to write tests or if you find them in playing with this new UI) in follow-up patches. Of course, if we discover something bad we can't fix easily, we won't ship it and we can back out it.
I'm a bit in a rush but I really understand you aren't and I don't want to rush you...
And I've few other things to implement... (telemetry and image settings mostly).
  
> One quickly observed bug, related to the asynchronicity of the code:
> 
>     1. Open the preview in [[Editor] Implement the new alt text flow (bug 1909604) #18492 (comment)](https://github.com/mozilla/pdf.js/pull/18492#issuecomment-2247947991).
> 
>     2. Add a new Stamp-editor.
> 
>     3. Immediately close the altText dialog _and_ delete the editor.
> 
> 
> In the console the following is printed (after a short wait):
> 
> ```
> TypeError: can't access property "guessedAltText", this[#currentEditor] is null
>     #mlGuessAltText new_alt_text_manager.js:248
> new_alt_text_manager.js:257:14
> ```

I fixed it two days ago.
I'll push a new version, with the enabled options in order to have a new viewer, and I'll set the options back, once we're more or less happy with patch.

calixteman commented:
/botio-linux preview

calixteman commented:
/botio-linux preview

calixteman commented:
@Snuffleupagus and FYI, QA already started to play with the preview and reported to me some bugs they found and I fixed. They'll continue to test it.

timvandermeij commented:
> Since we have an upcoming release (maybe this weekend), this should not land just before that anyway.

I'm going to make a new release this weekend, so that should be in time given that the soft-freeze for m-c is after the weekend. The advantage of releasing first is that we avoid any risks for the release because this is quite a big patch, and by merging this afterwards it'll have a month time to settle which gives us enough time to have all of this reviewed and any bugs we encounter afterwards fixed before the next release.

> Well, I hope you understand that I cannot review well over a thousand lines of new code immediately (and I'm still not at a 100% unfortunately).

Would it help if I do a pass over this PR during the weekend? I don't have extensive knowledge about the editor code, but perhaps an extra pair of eyes on the patch could help to increase confidence a bit if we need to land this in a short amount of time.

calixteman commented:
> Would it help if I do a pass over this PR during the weekend? I don't have extensive knowledge about the editor code, but perhaps an extra pair of eyes on the patch could help to reduce the review burden and increase confidence a bit if we need to land this in a short amount of time.

Any help is helpful ! and appreciated, thank you.

timvandermeij commented:
/botio-windows preview

timvandermeij commented on src/display/editor/editor.js:
```diff
@@ -1016,8 +1031,24 @@ class AnnotationEditor {
+  }
+
+  async setGuessedAltText(text) {
+    await this.#altText.setGuessedText(text);
```
It looks like `this.#altText` might be missing given the `#this.altText?` checks below, so shouldn't we also do that here and in the getter just above to make sure the access is safe?

Snuffleupagus commented on web/firefoxcom.js:
```diff
@@ -338,14 +344,32 @@ class MLManager {
+    if (enabled) {
+      await this.#loadAltTextEngine(false);
+    } else {
+      this.#enabled.delete(name);
```
Should we not remove it from `this.#ready` as well, to avoid inconsistent state?

Snuffleupagus reviewed (commented):
The implementation in `NewAltTextManager` feels a bit complicated, based on an initial look, but maybe it's not really possible to simplify it?

timvandermeij commented on src/display/editor/editor.js:
```diff
@@ -1016,8 +1031,24 @@ class AnnotationEditor {
+
   hasAltText() {
-    return !this.#altText?.isEmpty();
+    return this.#altText && !this.#altText.isEmpty();
```
If `this.#altText`  is `null` this function returns `null` while I'd expect this to always return a boolean. Should we use `!!this.altText` instead?

timvandermeij commented on web/app.js:
```diff
@@ -205,6 +206,14 @@ const PDFViewerApplication = {
       if (mode) {
         document.documentElement.classList.add(mode);
       }
+      // if (typeof PDFJSDev === "undefined" || PDFJSDev.test("TESTING")) {
```
Tracking comment to make sure the commented out code here and below gets replaced before merging.

timvandermeij commented on web/app_options.js:
```diff
@@ -187,9 +187,14 @@ const defaultOptions = {
   enableAltText: {
     /** @type {boolean} */
-    value: false,
+    value: true,
```
Tracking comment to make sure the changes in this file get updated before merging (it's also what makes linting fail now).

timvandermeij commented on web/genericcom.js:
```diff
@@ -56,9 +56,47 @@ class MLManager {
+    return false;
+  }
+
+  guess() {}
```
`guess(_data) {}` I assume, given that the `guess` method in `firefoxcom` has a `data` argument?

timvandermeij commented on web/new_alt_text_manager.js:
```diff
@@ -0,0 +1,396 @@
+/* Copyright 2023 Mozilla Foundation
```
s/2023/2024

timvandermeij commented on web/new_alt_text_manager.js:
```diff
@@ -0,0 +1,396 @@
+
+  #guessedAltText;
+
+  #isEditing = null;
```
I would expect a variable starting with `is` to be a boolean instead of `null`.

calixteman replied:
Yep I agree, but it was to be able to have something like `if (this.#isEditing !== value)...` the first time we set it.

timvandermeij commented on web/new_alt_text_manager.js:
```diff
@@ -0,0 +1,396 @@
+    this.#title = title;
+    this.#overlayManager = overlayManager;
+
+    dialog.addEventListener("close", this.#close.bind(this));
```
Should we use an `AbortController` to make sure that all event listeners get destroyed properly upon dialog close?

calixteman replied:
This ctor is called only one time and set the different listeners for the different elements.
So removing them when closing the dialog implies to add them again when we reopen it: it isn't what we are doing in the current alt_text_manager.js and I followed the same pattern.
That said, I'm not sure to know what's the best approach. But for sure we can do that in a follow-up.

timvandermeij commented on web/new_alt_text_manager.js:
```diff
@@ -0,0 +1,396 @@
+
+    this.#uiManager?.addEditListeners();
+    this.#currentEditor.altTextFinish();
+    this.#uiManager.setSelected(this.#currentEditor);
```
Shouldn't we use `this.#uiManager?.setSelected` here given the statement two lines above?

timvandermeij commented on web/viewer-geckoview.html:
```diff
@@ -65,6 +65,7 @@
           "display-node_utils": "../src/display/stubs.js",
 
           "web-alt_text_manager": "./stubs-geckoview.js",
+          "web-new_alt_text_manager": "./stubs-geckoview.js",
```
Let's move this down to keep the list ordered alphabetically for readability.

timvandermeij commented on web/viewer.html:
```diff
@@ -1,4 +1,4 @@
-﻿<!DOCTYPE html>
+<!DOCTYPE html>
```
We should revert this change. I assume it removed the UTF-8 BOM, but as mentioned in #18370 if we do that we should probably do it outside of this patch in case something breaks (we don't _expect_ it to, but it would be a shame if we'd have to fix it or back out this patch because of it).

timvandermeij reviewed (commented):
Looks good to me overall. The comments below are everything I've found going over the entire patch, but I'd like to see the updated patch and play with the preview a bit more before approving this.

calixteman commented on l10n/en-US/viewer.ftl:
```diff
@@ -416,3 +416,52 @@ pdfjs-editor-colorpicker-red =
+# This is a button that users can click to open the alt text editor and add alt text when it is not present.
+pdfjs-editor-new-alt-text-missing-button-label = Missing alt text
+
+# This is a button that opens up the alt text modal where users should review the alt text that was automatically generated.
```
It has been written by an American.

calixteman commented on src/display/editor/editor.js:
```diff
@@ -223,6 +226,15 @@ class AnnotationEditor {
       ])
     );
+
+    AnnotationEditor._l10nPromise.set(
```
Because the corresponding string has a parameter, so we need to call the got function with the parameter in order to get the translated string.

calixteman commented on src/display/editor/editor.js:
```diff
@@ -223,6 +226,15 @@ class AnnotationEditor {
         l10n.get(str.replaceAll(/([A-Z])/g, c => `-${c.toLowerCase()}`)),
       ])
     );
+
+    AnnotationEditor._l10nPromise.set(
+      "pdfjs-editor-new-alt-text-generated-alt-text-with-disclaimer",
+      l10n.get.bind(
+        l10n,
+        "pdfjs-editor-new-alt-text-generated-alt-text-with-disclaimer"
+      )
+    );
+
```
Because the corresponding string has a parameter, so we need to call the got function with the parameter in order to get the translated string.

timvandermeij commented:
/botio-linux preview

timvandermeij commented on web/new_alt_text_manager.js:
```diff
@@ -0,0 +1,395 @@
+    this.#uiManager = null;
+  }
+
+  #save() {
```
If the PDF with image and alt text is saved to disk and opened again in PDF.js I don't see the alt text in the DOM. I would assume it to end up somewhere so screen readers can find it. Could you explain how that works?

calixteman replied:
The alt-text is in the struct tree under the `canvas`:
![image](https://github.com/user-attachments/assets/3f04f2a1-bc75-4a72-9e80-80f5effa957b)
That said maybe we should just add the alt-text as a aria-*** on the annotation itself.

timvandermeij reviewed (approved):
Looks good to me, with:

- the comments below and the two open tracking comments above addressed;
- passing CI and tests;
- a follow-up issue to create integration tests for the new alt text flow because currently we only test the current/old flow (note that we can wait with that until the new alt text flow is accepted and it's combined with removing the current/old flow).

calixteman commented:
/botio integrationtest

calixteman commented:
/botio integrationtest

calixteman commented:
@Snuffleupagus is there anything blocking ?
